### PR TITLE
Create functionality to store currentPage in the history

### LINKED ISSUE #4976: Returning from resource details view resets page number
* A user navigates to the resources view and requests the details of a resource present on a page larger than 1
* When the user clicks the back arrow from the resource details view, it is routed back to the first page of the resources view and not to the page the user came from.

[Screencast from 2023-07-18 07-44-42.webm](https://github.com/inmanta/web-console/assets/2684622/5008ebc7-1682-4850-a1ce-7d31b29b57f2)

### PR DESCRIPTION
**Pull request by matborowczyk**
# Description

* Short description here *

closes #4976 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Code is clear and sufficiently documented
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )

### CODE REVIEW

File: src/Data/Common/UrlState/useUrlStateWithCurrentPage.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import { isEqual, isArray } from "lodash";
+import { provide, Location, Replace, StateConfig, Update } from "./helpers";
+import { handleUrlState } from "./useUrlState";
+
+export const useUrlStateWithCurrentPage = provide(
```
Comment (LukasStordeur) [inline src/Data/Common/UrlState/useUrlStateWithCurrentPage.ts:5]: Can you add comments to this, about the logic and usage?

File: src/Data/Common/UrlState/useUrlStateWithCurrentPage.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  handleUrlStateWithCurrentPage,
+);
+const valueIsValid = (value: unknown): value is string[] => {
```
Comment (LukasStordeur) [inline src/Data/Common/UrlState/useUrlStateWithCurrentPage.ts:8]:
```suggestion
/**
* Checks whether the value passed is an Array of strings. 
* @returns boolean
*/
const valueIsValid = (value: unknown): value is string[] => {
```

File: src/Data/Common/UrlState/useUrlStateWithCurrentPage.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  return isArray(value) && value.every((param) => typeof param === "string");
+};
+
+const from = (value: string[]): CurrentPage => {
+  return { kind: "CurrentPage", value: valueIsValid(value) ? value : [] };
+};
+const serialize = (currentPage: CurrentPage): string[] => {
```
Comment (LukasStordeur) [inline src/Data/Common/UrlState/useUrlStateWithCurrentPage.ts:15]: serialize is for me more than just returning the currentPage.value.

File: src/Data/Common/UrlState/useUrlStateWithCurrentPage.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  return currentPage.value;
+};
+
+const parse = (candidate: unknown): CurrentPage | undefined => {
+  return valueIsValid(candidate) ? from(candidate) : undefined;
+};
+
+export interface CurrentPage {
+  kind: "CurrentPage";
+  value: string[];
```
Comment (LukasStordeur) [inline src/Data/Common/UrlState/useUrlStateWithCurrentPage.ts:25]: Why did you go for the array of strings?

File: src/Data/Managers/GetInstanceResources/QueryManager.ts [hunk 5]
```diff
@@ -159,6 +159,7 @@ export function InstanceResourcesQueryManager(
         kind: "GetServiceInstances",
         name: serviceEntity,
         pageSize: PageSize.initial,
+        currentPage: { kind: "CurrentPage", value: [] },
```
Comment (LukasStordeur) [inline src/Data/Managers/GetInstanceResources/QueryManager.ts:162]: why are you not using the InitialCurrentPage here?

File: src/Data/Managers/ServiceInstances/QueryManager.test.tsx [hunk 6]
```diff
@@ -41,27 +43,37 @@ const setup = () => {
...
       name: "name",
       pageSize: PageSize.initial,
+      currentPage: currentPageMock,
     });
 
     if (!RemoteData.isSuccess(data)) return null;
 
     const onNext = () => {
-      data.value.handlers.next && data.value.handlers.next();
+      console.log(data.value.handlers.next);
```
Comment (LukasStordeur) [inline src/Data/Managers/ServiceInstances/QueryManager.test.tsx:59]: I believe this one can be removed.

File: src/Data/Managers/ServiceInstances/getUrl.test.ts [hunk 7]
```diff
@@ -64,6 +67,7 @@ test("getUrl returns correct url for sort & filter", () => {
       order: "asc",
     },
     pageSize: PageSize.initial,
+    currentPage: { kind: "CurrentPage", value: [] },
```
Comment (LukasStordeur) [inline src/Data/Managers/ServiceInstances/getUrl.test.ts:70]: Maybe use the initialCurrentPage ?

File: src/Slices/CreateInstance/UI/CreateInstance.test.tsx [hunk 8]
```diff
@@ -204,7 +204,7 @@ test("Given the CreateInstance View When creating an instance with Inter-service
   });
   expect(apiHelper.pendingRequests[0]).toEqual({
     method: "GET",
-    url: `/lsm/v1/service_inventory/${InterServiceRelations.editable.entity_type}?include_deployment_progress=False&limit=100&filter.order_id=a`,
+    url: `/lsm/v1/service_inventory/${InterServiceRelations.editable.entity_type}?include_deployment_progress=False&limit=250&filter.order_id=a`,
```
Comment (LukasStordeur) [inline src/Slices/CreateInstance/UI/CreateInstance.test.tsx:207]: why did you change the default here to 250?

File: src/Slices/DesiredState/UI/Page.tsx [hunk 9]
```diff
@@ -22,6 +23,10 @@ export const Page: React.FC = () => {
   const [versionToDelete, setVersionToDelete] = useState<ParsedNumber>(0);
   const [errorMessage, setErrorMessage] = useState("");
   const { queryResolver } = useContext(DependencyContext);
+
+  const [currentPage, setCurrentPage] = useUrlStateWithCurrentPage({
+    route: "CompileReports",
```
Comment (LukasStordeur) [inline src/Slices/DesiredState/UI/Page.tsx:28]: Aren't you on the DesiredState page? Why do you need the CompileReports ?

File: src/Slices/DesiredStateDetails/Data/getUrl.spec.ts [hunk 10]
```diff
@@ -14,6 +14,7 @@ it.each`
       version: "123",
       pageSize: PageSize.from(pageSize),
       filter,
+      currentPage: { kind: "CurrentPage", value: [] },
```
Comment (LukasStordeur) [inline src/Slices/DesiredStateDetails/Data/getUrl.spec.ts:17]: initialCurrentPage ?

File: src/Slices/Events/Data/getUrl.test.ts [hunk 11]
```diff
@@ -26,6 +26,7 @@ describe("getUrl for events ", () => {
         filter,
         sort,
         pageSize: PageSize.from(size),
+        currentPage: { kind: "CurrentPage", value: [] },
```
Comment (LukasStordeur) [inline src/Slices/Events/Data/getUrl.test.ts:29]: initialCurrentPage ?

File: src/Slices/Facts/Data/getUrl.test.ts [hunk 12]
```diff
@@ -17,6 +17,7 @@ describe("getUrl for facts", () => {
         pageSize: PageSize.from(size),
         filter,
         sort,
+        currentPage: { kind: "CurrentPage", value: [] },
```
Comment (LukasStordeur) [inline src/Slices/Facts/Data/getUrl.test.ts:20]: initialCurrentPage ?

File: src/Slices/Notification/Core/Query.ts [hunk 13]
```diff
@@ -43,6 +45,7 @@ export const drawerQuery: Query = {
   origin: "drawer",
   pageSize: PageSize.from("100"),
   filter: { cleared: false },
+  currentPage: { kind: "CurrentPage", value: [] },
```
Comment (LukasStordeur) [inline src/Slices/Notification/Core/Query.ts:48]: initialCurrentPage ?

File: src/Slices/Parameters/Data/getUrl.test.ts [hunk 14]
```diff
@@ -17,6 +17,7 @@ it.each`
       pageSize: PageSize.from(pageSize),
       filter,
       sort,
+      currentPage: { kind: "CurrentPage", value: [] },
```
Comment (LukasStordeur) [inline src/Slices/Parameters/Data/getUrl.test.ts:20]: initialCurrentPage ?

File: src/Slices/Resource/Data/getUrl.test.ts [hunk 15]
```diff
@@ -17,6 +17,7 @@ describe("getUrl for latest released resources ", () => {
         pageSize: PageSize.from(size),
         filter,
         sort,
+        currentPage: { kind: "CurrentPage", value: [] },
```
Comment (LukasStordeur) [inline src/Slices/Resource/Data/getUrl.test.ts:20]: initialCurrentPage ?

File: src/Slices/Resource/Data/getUrl.ts [hunk 16]
```diff
@@ -21,5 +22,9 @@ export function getUrl({
         )}`
       : "";
   const sortParam = sort ? `&sort=${sort.name}.${sort.order}` : "";
-  return `/api/v2/resource?deploy_summary=True&limit=${pageSize.value}${filterParam}${sortParam}`;
+  const currentPageParams =
```
Comment (LukasStordeur) [inline src/Slices/Resource/Data/getUrl.ts:25]: I believe this is something worth extracting into your handler, and exported. That way, if the logic needs to be adjusted a bit, you'll only need to change it in one place, instead of all 50 files where this is being used ;)

File: src/Slices/ResourceDetails/Data/ResourceHistory/getUrl.test.ts [hunk 17]
```diff
@@ -15,6 +15,7 @@ it.each`
       id: resourceId,
       pageSize: PageSize.from(pageSize),
       sort,
+      currentPage: { kind: "CurrentPage", value: [] },
```
Comment (LukasStordeur) [inline src/Slices/ResourceDetails/Data/ResourceHistory/getUrl.test.ts:18]: initialCurrentPage ?

File: src/Slices/ResourceDetails/Data/ResourceLogs/getUrl.spec.ts [hunk 18]
```diff
@@ -17,6 +17,7 @@ it.each`
       id: resourceId,
       pageSize: PageSize.from(pageSize),
       filter,
+      currentPage: { kind: "CurrentPage", value: [] },
```
Comment (LukasStordeur) [inline src/Slices/ResourceDetails/Data/ResourceLogs/getUrl.spec.ts:20]: initialCurrentPage ?

File: src/UI/Components/ServiceInstanceForm/Components/AutoCompleteInputProvider.tsx [hunk 19]
```diff
@@ -35,7 +36,8 @@ export const AutoCompleteInputProvider: React.FC<Props> = ({
     kind: "GetServiceInstances",
     name: serviceName,
     filter,
-    pageSize: PageSize.from("100"),
+    pageSize: PageSize.from("250"),
```
Comment (LukasStordeur) [inline src/UI/Components/ServiceInstanceForm/Components/AutoCompleteInputProvider.tsx:39]: I don't think it's wise to change it by default to 250 (remember the performance issues from our last teambuilding. That page would display 150 more rows of loaded data, by default.)

**Review by LukasStordeur — COMMENTED** [review]
I haven't seen any update in the E2E tests, seen the changes, I think it's best to have at least one page where you navigate back and forth, and assert that the page has been retained. 

did you also update the text being displayed in the dropdown? I might have missed it in the PR.

File: src/Data/Common/UrlState/useUrlStateWithCurrentPage.ts (hunk 4 above)
Comment (matborowczyk) [reply src/Data/Common/UrlState/useUrlStateWithCurrentPage.ts:25]: There are Two sets of two arguments responsible for dictating on which page we are(depending if we are going froward or backward), separated by "&" so I thought that if they are always in pairs, then instinctively picked this one.

File: src/UI/Components/ServiceInstanceForm/Components/AutoCompleteInputProvider.tsx (hunk 19 above)
Comment (matborowczyk) [reply src/UI/Components/ServiceInstanceForm/Components/AutoCompleteInputProvider.tsx:39]: I understood that component as an autocomplete for i.e. relations so I thought it's fine, but I can revert it

File: src/Data/Managers/GetInstanceResources/QueryManager.ts (hunk 5 above)
Comment (matborowczyk) [reply src/Data/Managers/GetInstanceResources/QueryManager.ts:162]: Had to miss those, probably to wrong search formula, as I added initialCurrentPage after all those changes

File: src/Slices/CreateInstance/UI/CreateInstance.test.tsx (hunk 8 above)
Comment (matborowczyk) [reply src/Slices/CreateInstance/UI/CreateInstance.test.tsx:207]: This request is for AutoComplete functionality, it was using the max range available, to serve the most possibilities  to autocomplete so I thought that it should be increased from previous max to current

File: src/Data/Common/UrlState/useUrlStateWithCurrentPage.ts (hunk 3 above)
Comment (matborowczyk) [reply src/Data/Common/UrlState/useUrlStateWithCurrentPage.ts:27]: rearranged a little bit, but still that's the most of serialisation  that could be done inside

File: src/Data/Common/UrlState/useUrlStateWithCurrentPage.ts (hunk 1 above)
Comment (matborowczyk) [reply src/Data/Common/UrlState/useUrlStateWithCurrentPage.ts:9]: Added about the wrapper, and if someone will go to this file then also will get the comment bellow about the actual handler of CurrentPage state

**Review by LukasStordeur — APPROVED** [review]
